pkg/controller/deployment/util/deployment_util.go

```diff
@@ -495,19 +495,24 @@ func getReplicaSetFraction(logger klog.Logger, rs apps.ReplicaSet, d apps.Deploy
 		return -*(rs.Spec.Replicas)
 	}
 
-	deploymentReplicas := *(d.Spec.Replicas) + MaxSurge(d)
-	annotatedReplicas, ok := getMaxReplicasAnnotation(logger, &rs)
-	if !ok {
+	deploymentMaxReplicas := *(d.Spec.Replicas) + MaxSurge(d)
+	deploymentMaxReplicasBeforeScale, ok := getMaxReplicasAnnotation(logger, &rs)
+	if !ok || deploymentMaxReplicasBeforeScale == 0 {
 		// If we cannot find the annotation then fallback to the current deployment size. Note that this
 		// will not be an accurate proportion estimation in case other replica sets have different values
 		// which means that the deployment was scaled at some point but we at least will stay in limits
-		// due to the min-max comparisons in getProportion.
-		annotatedReplicas = d.Status.Replicas
+		// due to the min-max comparisons in GetReplicaSetProportion.
+		deploymentMaxReplicasBeforeScale = d.Status.Replicas
+		if deploymentMaxReplicasBeforeScale == 0 {
+			// Rare situation: missing annotation and pods are failing to be created.
+			return 0
```

Comment:
This place (out of 2) is the only change in this function. The rest are renames.

Reply:
Similarly, looks like a godd idea to safeguard the code, but maybe this is also a fix for potential panic which could exist in e2e? I guess if this is possible, then the result of the function is unexpected as we skip the scaling, so it is probably worth leaving a log at V(2).

Reply:
As mentioned in https://github.com/kubernetes/kubernetes/pull/128535#discussion_r1827737545 the issue is not that big. And we do log the incorrect annotations (also improved here: https://github.com/kubernetes/kubernetes/pull/128527).